Spawn subprocess

granders opened the pull request:
PR into parallel-test-runner branch since I plan to do a few more incremental pull requests before merging parallel-test-runner into master.

Tests still run sequentially, but this PR updates ducktape to run tests in subprocesses, using message passing to coordinate starting of new tests.

granders commented:
@ewencp @hachikuji 
Would you mind doing an initial once-over?

ConfluentJenkins commented:
Test FAILed.

ConfluentJenkins commented:
Test PASSed.

ewencp commented on ducktape/command_line/main.py:
```diff
@@ -158,18 +159,19 @@ def main():
 
     # Initializing the cluster is slow, so do so only if
     # tests are sure to be run
+    cluster = None
```
Why this assignment? Python isn't scoped the same way as other languages, and if an exception occurs you're going to exit anyway.

ewencp commented on ducktape/tests/loader.py:
```diff
@@ -33,25 +33,31 @@ class LoaderException(Exception):
         :raise LoaderException if it can't be parsed
+
+        Examples:
+            "path/to/directory" -> ("/absolute/path/to/directory", "", "", "")
```
Is the `/absolute/path/to/directory` part of this accurate anymore? You don't have an `abspath` call and `os.path.dirname` does not guarantee an absolute path.

ewencp commented on ducktape/tests/loader.py:
```diff
@@ -141,7 +173,12 @@ def import_modules(self, file_list):
                 try:
-                    module_list.append(importlib.import_module(module_name))
+                    module_list.append(
+                        {
```
Given how this is actually used:

```
        modules_and_files = self.import_modules(test_files)
        for mf in modules_and_files:
            test_context_list.extend(self.expand_module(mf))
```

it might be better to just use tuples, which admittedly leave a bit of the info implicit, but can be clearer and less awkward:

```
        for module,file in self.import_modules(test_files):
            test_context_list.extend(self.expand_module(module, file))
```

ewencp commented on ducktape/tests/loader.py:
```diff
@@ -192,16 +229,24 @@ def import_modules(self, file_list):
         return module_list
 
-    def expand_module(self, module):
+    def expand_module(self, module_and_file):
```
This and the dict are kind of awkward. Why not choose either a concrete named type (with expected structure) or separate this into two separate parameters?

It might also help to rename some of these methods to have a `_` prefix so we know when they are internal (and therefore it might not be *as* bad for them to have somewhat weird arguments like this one).

ewencp commented on ducktape/tests/message.py:
```diff
@@ -0,0 +1,95 @@
+    FINISHED = "FINISHED"
+    LOG = "LOG"
+
+    def __init__(self, source_id):
```
I'm not sure I understand what this class is supposed to represent. From its name I thought it had to do with requests being sent between the different processes, but it seems this object only has a `source_id` and each of the methods seems to generate an `event` more like what I was expecting to be sent back and forth?

ewencp commented on ducktape/tests/message.py:
```diff
@@ -0,0 +1,95 @@
+        assert len(set(event.keys()).intersection(set(payload.keys()))) == 0, \
+            "Payload and base event should not share keys. base event: %s, payload: %s" % (str(event), str(payload))
+
+        event. update(payload)
```
nit: stray space

ewencp commented on ducktape/tests/result.py:
```diff
@@ -34,41 +33,40 @@ def __str__(self):
 
-        self.test_context = test_context
-        self.session_context = self.test_context.session_context
+        self.test_id = test_context.test_id
```
Is all this copying to make this serializable or something? Seems like we're just extracting everything out of `test_context`.

granders replied:
yeah - Test context isn't currently serializable, so we can't have a reference to a test_context object if we want to serialize TestResult

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,212 +14,175 @@
+    def __init__(self, cluster, session_context, session_logger, tests, port=ConsoleDefaults.TEST_DRIVER_PORT):
+        # session_logger, message logger,
+        self.session_logger = session_logger
+        self.request_logger = RequestLogger(session_context).logger
```
Any reason not to just share the session logger?

granders replied:
Not necessarily; it's more a question of user interaction
i.e. is it useful for debugging to have a separate `request_log` file? Maybe maybe not

ewencp replied:
It just seemed like we were already putting logs from the test runners into the session/test logger anyway. I figured the number of requests here should be minimal and mostly just bookend the test, right?

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,212 +14,175 @@
+        elif event["event_type"] == Request.LOG:
+            self.handle_log(event)
+        else:
+            self.handle_other(event)
```
Shouldn't this just be an error since it is unexpected?

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,212 +14,175 @@
-            if self.stop_testing:
-                break
+    def handle_log(self, event):
+        self.log(event["log_level"], event["message"])
```
If you don't do the ack message immediately, you probably want to protect against exceptions in this code. This bit isn't that bad, but you might want to standardize the pattern if you want to just always ack first.

Also, is this log missing some information about who is running it, i.e. the `source_id` I see elsewhere? Seems like that is how you plan to track the active test state properly when you have parallel tests (and presumably also the lifecycle events, which don't seem to do anything right now).

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,212 +14,175 @@
-                break
+    def handle_log(self, event):
+        self.log(event["log_level"], event["message"])
+        self.socket.send(self.serde.serialize(event))
```
Why are we sending the event back in reply? Looks like we do this repeatedly. Why wouldn't this just be empty or some simple ack?

granders replied:
Good call - this was intended to be very temporary - it was just a convenient way to ack without extra effort

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+
+    def __init__(self, port, logger_name, log_dir, debug, max_parallel):
+        self.serde = SerDe()
+        self.logger = TestLogger(logger_name, log_dir, debug, max_parallel).logger
```
Why would `max_parallel` ever be a relevant parameter to the `TestLogger`? In the code it looks like maybe you are using this as a proxy for when you want to log debug level to stdout. Shouldn't this just use the `--debug` CLI flag (which may need to be passed through a couple of layers to get here)?

granders replied:
The question I was thinking about here is what should the behavior be if debug is enabled when running tests in parallel?

- should we let the test runner client subprocesses compete for lock on stdout and dump all the interleaved messages?
- should we just print session log debug info? This is the less likely use case and tries to keep some sanity to the output.

Either way, the most likely use case for wanting the debug output is when you're debugging your own test(s), and running test(s) serially

ewencp replied:
Not sure I even understand the second bullet point. Loggers are thread safe anyway, right? The default behavior will just be the first bullet. In any case, I'm confused how this relates to the `max_parallel` parameter?

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+
+        self.id = "test-runner-%d-%d" % (os.getpid(), id(self))
+        self.request = Request(self.id)
+        self.sender = Sender("localhost", str(self.runner_port), self.logger)
```
Shouldn't hostname be passed through the constructor/`run_client` if you eventually want this running on a different host?

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+    def collect_test_context(self, directory, file_name, cls_name, method_name, injected_args):
+        # TODO - different logger for TestLoader object
+        loader = TestLoader(self.session_context, self.logger, injected_args=injected_args, cluster=self.cluster)
+        loaded_context_list = loader._discover(directory, file_name, cls_name, method_name)
```
`_discover` seems misnamed since you're using it from another class?

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+        self.test_context = self.collect_test_context(**self.test_metadata)
+
+        self.send(self.request.running())
+        if len(self.cluster) != self.cluster.num_available_nodes():
```
Strictly speaking I don't think this is always true. `LocalhostCluster` initializes the available nodes to `sys.maxint` and also always returns `sys.maxint` for `len`. JsonCluster uses its explicit list to keep everything in sync.

We may just not have clearly defined the semantics, but I don't think this test will always work. A `DockerCluster` might be another case similar to `LocalhostCluster` where there isn't a fixed size up front, and it could effectively have an infinitely large cluster.

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+        self.send(self.request.running())
+        if len(self.cluster) != self.cluster.num_available_nodes():
+            # Sanity check - are we leaking cluster nodes?
+            raise RuntimeError(
```
Should this be doing anything to notify the parent process with some `self.send(something)`?

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+            test_reporter = SingleResultFileReporter(result)
+            test_reporter.report()
+
+        # Tell the server we are finished
```
Again, in the case of test failure we would have hit the above `except` and `finally`, but we're not going to process this bit of code. Intended or is this going to cause the parent process to hang?

granders replied:
yep it'll cause the parent process to hang :)
I plan to add a "liveness" check ("is the subprocess pid still present?") in the event loop. Even with such a check, the parent process will still hang if the test itself hangs and the subprocess never exits.

It's interesting that even to handle the non-parallel case, a liveness check on the subprocesses can help handle certain failures cleanly.

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+            self.logger.log(log_level, msg)
+            self.test.logger.log(log_level, msg)
+
+        self.send(self.request.log(msg, level=log_level))
```
I thought we were just going to be forwarding logs, but now it seems like we potentially log to 3 separate places? Why so many?

granders replied:
Oops, that' should've been two places (I'll delete `self.test.logger.log(log_level, msg)` since `self.test.logger` and `self.logger` are the same object)

Two places continues the previous logging behavior, where notable events from the "thing running the test" are propagated both the the session_log and the test_log

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,281 @@
+                    self.close()
+                    self._init_socket()
+                    waiting_for_reply = False
+                time.sleep(.5)
```
Why `time.sleep`? Why not just block indefinitely or set `REQUEST_TIMEOUT_MS` much, much larger? It looks like maybe zmq doesn't do anything in terms of reliability by default. This is pretty concerning... Knowing that failures can occur is good, and having a way to actually handle them gracefully is better, but firing and praying doesn't seem great, especially since I'm not sure all the code on the other half of this connection is idempotent. In fact, the logging code *definitely* isn't idempotent.

granders replied:
good point, will ditch the sleep

I'm not sure we want a big timeout here while we're still running everything on the same machine, or even very many retries. Lack of a response when everything is running on the same machine most likely suggests that the parent process died?

Idempotence of the "server" logic is a good point, but practically speaking, it's not clear to me that this is a case we need to handle gracefully? Given that everything is on the same machine, on the server side I wonder if it's reasonable to treat multiple occurrences of certain events from from the same client client as a fatal error and leave it at that.

ewencp replied:
My concern is more that we're starting with a networking library that doesn't really offer us the semantics we want. For example, why aren't we just using some TCP sockets with some reasonable timeouts and getting nice reliability guarantees. If we have socket disconnections and we expect reliable intra-DC connections to be the norm, it's fine to not handle them particularly gracefully (e.g. kill the test or even the entire test run).

ewencp commented on ducktape/tests/test.py:
```diff
@@ -173,20 +173,74 @@ def _escape_pathname(s):
+
+    @property
+    def logger_name(self):
+        return self._logger_name
```
When there's no computation going on, is there any reason we're not just setting it as the field name and not including these extra lines of code for wrapping that field?

granders replied:
ack

ewencp commented on ducktape/tests/test.py:
```diff
@@ -292,41 +360,10 @@ def test_name(self):
-
+    def logger(self):
+        """Lazily instantiate logger"""
+        return TestLogger(
```
It looks like all uses of `TestLogger` are like this -- you construct the class, then immediately just take the `logger` property. That makes me think it should probably be a function, not a class...

ewencp commented on setup.py:
```diff
@@ -42,7 +42,7 @@ def run_tests(self):
       packages=find_packages(),
       package_data={'ducktape': ['templates/report/*']},
-      install_requires=['jinja2', 'requests', 'pysistence'],
+      install_requires=['jinja2', 'requests', 'pysistence', 'zmq'],
```
The `zmq` package seems to suggest you should be using `pyzmq`: https://pypi.python.org/pypi/zmq/0.0.0

ewencp commented on tests/runner/check_serial_runner.py:
```diff
@@ -12,29 +12,55 @@
-from ducktape.tests.runner import SerialTestRunner
-from ducktape.mark import ignore, parametrize
+from ducktape.tests.test import TestContext
+from ducktape.tests.runner import TestRunner
```
filename should probably change given that `SerialTestRunner` no longer exists?

ewencp commented on tests/tests/check_session.py:
```diff
@@ -12,16 +12,40 @@
+        kwargs = {
+            "session_id": "hello-123",
+            "results_dir": self.tempdir,
+            "cluster": LocalhostCluster(),
```
Only issue with this test may be that `LocalhostCluster` is basically the most trivial implementation. Not sure we'll be confident that others would work based on this test.

ConfluentJenkins commented:
Test PASSed.

ConfluentJenkins commented:
Test PASSed.

ConfluentJenkins commented:
Test PASSed.

granders commented:
@ewencp I addressed most of your comments, and added a couple JIRAs for:
- handling duplicate messages from client
- prevent runner hang if runner client dies

ewencp commented on ducktape/tests/event.py:
```diff
@@ -0,0 +1,150 @@
+import time
+
+
+class ClientEventSupplier(object):
```
not sure about in python, but Factory is probably the more common idiom than Supplier

ewencp commented on ducktape/tests/event.py:
```diff
@@ -0,0 +1,150 @@
+        self.source_id = source_id
+        self.event_id = 0
+
+    def event(self, event_type, payload=None):
```
is this meant to be public or should it be named `_event`

ewencp commented on ducktape/tests/loader.py:
```diff
@@ -26,32 +27,103 @@ class LoaderException(Exception):
 
 
+# A helper container class
+ModuleAndFile = collections.namedtuple('ModuleAndFile', 'module file')
```
Interesting, I didn't know you could do `'module file'` instead of `['module', 'file']`

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,216 +14,153 @@
+        self.zmq_context = zmq.Context()
+        self.socket = self.zmq_context.socket(zmq.REP)
+        self.socket.bind("tcp://*:%s" % str(self.port))
+        # self.poller = zmq.Poller()
```
Stray commented code, probably should just be deleted

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,216 +14,153 @@
+    def handle(self, event):
+        self.log(logging.DEBUG, str(event))
+
+        if event["event_type"] == ClientEventSupplier.READY:
```
If the number of events stays small here, this approach works fine and is perfectly readable. If this grows substantially, you might just want to use a dictionary-based dispatch, i.e. you build a static dictionary like:

```
event_handlers = { ClientEventSupplier.READY: self.handle_ready, .... }
```

And then dispatch is just:

```
self.event_handlers[event["event_type"]](event)
```

ewencp commented on ducktape/tests/runner.py:
```diff
@@ -14,216 +14,153 @@
+        """Log an event received from the client."""
+        pass
+
+    def log(self, log_level, msg):
```
Probably doesn't matter here as its just for this class, but if we're proxying for loggers, you might want to provide `*args` and `**kwargs` support as in the `logging` module (https://docs.python.org/3/library/logging.html#logging.Logger.log) since it doesn't require you to do the interpolation step yourself.

ewencp commented on ducktape/tests/runner_client.py:
```diff
@@ -0,0 +1,283 @@
+    def send(self, event):
+        return self.sender.send(event)
+
+    def collect_test_context(self, directory, file_name, cls_name, method_name, injected_args):
```
should probably be private. probably same for `send` above too, as well as some other methods below -- almost none of this class needs to be public. (would probably be good to start following good practice naming private/public fields as well)

ewencp commented on ducktape/tests/test.py:
```diff
@@ -173,20 +173,79 @@ def _escape_pathname(s):
+
+        ch = logging.StreamHandler(sys.stdout)
+        ch.setFormatter(formatter)
+        if self.debug and self.max_parallel == 1:
```
Still concerned about why we're doing this. If the user asks for debug, seems like they should get debug level output... `max_parallel` shouldn't have any impact on their request.

granders replied:
Ack, ok that's reasonable

ewencp commented on ducktape/tests/test.py:
```diff
@@ -292,41 +365,7 @@ def test_name(self):
-
+    def logger(self):
+        """Lazily instantiate logger"""
+        return test_logger(
```
This isn't just lazy, it actually reinstantiates if on every call. I suspect you want to cache this once it is constructed.

ewencp commented:
@granders A bunch more minor cleanup comments, but LGTM

ConfluentJenkins commented:
Test PASSed.

granders commented:
Thanks for the feedback @ewencp
